tests/store/tracking/test_rest_store.py

```diff
@@ -1690,3 +1693,155 @@ def test_delete_scorer_without_version():
         mock_call_endpoint.assert_called_once_with(
             DeleteScorer, message_to_json(DeleteScorer(experiment_id=experiment_id, name=name))
         )
+
+
+def _create_mock_response(status_code: int = 200, text: str = "{}") -> mock.MagicMock:
+    """Helper to create a mock HTTP response."""
+    response = mock.MagicMock()
+    response.status_code = status_code
+    response.text = text
+    return response
+
+
+def _create_test_spans() -> list[LiveSpan]:
+    """Helper to create test spans for log_spans tests."""
+    otel_span = create_mock_otel_span(
+        trace_id=123,
+        span_id=1,
+        name="test_span",
+        start_time=1000000,
+        end_time=2000000,
+    )
+    return [LiveSpan(otel_span, trace_id="tr-123")]
+
+
+def test_log_spans_with_version_check():
+    """Test that log_spans raises NotImplementedError for old server versions."""
+    spans = _create_test_spans()
+    experiment_id = "exp-123"
+
+    # Test 1: Server version is None (failed to retrieve)
+    # Use unique host to avoid cache conflicts
+    creds1 = MlflowHostCreds("https://host1")
+    store1 = RestStore(lambda: creds1)
+    with mock.patch("mlflow.store.tracking.rest_store.http_request") as mock_http:
+        mock_http.side_effect = Exception("Connection error")
+        with pytest.raises(NotImplementedError, match="could not identify MLflow server version"):
+            store1.log_spans(experiment_id, spans)
+
+    # Test 2: Server version is less than 3.4
+    creds2 = MlflowHostCreds("https://host2")
+    store2 = RestStore(lambda: creds2)
+    with mock.patch("mlflow.store.tracking.rest_store.http_request") as mock_http:
+        mock_http.return_value = _create_mock_response(text="3.3.0")
+        with pytest.raises(
+            NotImplementedError, match="MLflow server version 3.3.0 is less than 3.4"
+        ):
+            store2.log_spans(experiment_id, spans)
+
+    # Test 3: Server version is exactly 3.4.0 - should succeed
+    creds3 = MlflowHostCreds("https://host3")
+    store3 = RestStore(lambda: creds3)
+    with mock.patch("mlflow.store.tracking.rest_store.http_request") as mock_http:
+        # First call is to /version, second is to OTLP endpoint
+        mock_http.side_effect = [
+            _create_mock_response(text="3.4.0"),  # version response
+            _create_mock_response(),  # OTLP response
+        ]
+        result = store3.log_spans(experiment_id, spans)
+        assert result == spans
```

Comment:
```suggestion
    with mock.patch("mlflow.store.tracking.rest_store.http_request", side_effect=[
            _create_mock_response(text="3.4.0"),  # version response
            _create_mock_response(),  # OTLP response
        ]) as mock_http:
        result = store3.log_spans(experiment_id, spans)
        assert result == spans
```

we can pass `side_effect` when calling  `mock.patch`.